python/ray/tune/durable_trainable.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+from ray.tune.trainable import Trainable
+from ray.tune.syncer import get_cloud_sync_client
+
+
+class DurableTrainable(Trainable):
+    """A fault-tolerant Trainable.
+
+    Supports checkpointing to and restoring from remote storage.
+
+    The storage client must provide durability for restoration to work. That
+    is, once ``storage.client.wait()`` returns after a checkpoint `sync up`,
+    the checkpoint is considered committed and can be used to restore the
+    trainable.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DurableTrainable, self).__init__(*args, **kwargs)
+        self.storage_client = get_cloud_sync_client(self._root_storage_path())
+
+    def save(self, checkpoint_dir=None):
```

Comment:
For this, can we just enforce that `save` will use `checkpoint_dir`, and we can just sync the entire checkpoint_dir rather than need to consider some arbitrary subpath?

Reply:
added a check that the dir has to be `logdir`, or within. i think syncing the individual checkpoint dirs should be fine otherwise though.